File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {

        vector<vector<string>> result;

        // strs = ["eat","tea","tan","ate","nat","bat"]
        // Output: [["bat"],["nat","tan"],["ate","eat","tea"]]

        // We can have a map -> sorted(key) : vector<val>

        unordered_map<string, vector<string>> groups;

        for(int i = 0; i < strs.size(); i++){

            string word = strs[i];
            sort(word.begin(), word.end());

            groups[word].push_back(strs[i]);
        }

        for(const auto & pair : groups){
            result.push_back(pair.second);
        }

        return result;
        
    }
};

File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:

        # eat tea tan ate nat bat
        # sort them out we can create a dictionary and store values of it

        # my_dict = {}
        my_dict = defaultdict(list)

        for word in strs:
            # key = sorted(word)
            key = "".join(sorted(word))
            my_dict[key].append(word)

        
        # answer: list[list[str]] = []
        answer = []


        for key in my_dict:
            answer.append(my_dict.get(key))
        
        return answer

        